### PR TITLE
Added support to print as alternate platform specific event data

### PR DESCRIPTION
**Pull request by dtacalau**
This has helped me during debugging windows issues. Was thinking it's a good idea to have it on master too.

### CODE REVIEW

**Comment by Thomasdezeeuw** [comment]
I don't think any types in libc implement `fmt::Debug` so that is why the CI is failing on Unix platforms.

**Comment by dtacalau** [comment]
> I don't think any types in libc implement `fmt::Debug` so that is why the CI is failing on Unix platforms.

@Thomasdezeeuw Got it, I limited this change to Windows only.

**Comment by Thomasdezeeuw** [comment]
I was thinking about this a bit and let's do this only when providing `{:#?}`, normally this means the output is pretty printed, but we can (ab)use to also add additional platforms specific information. For Unix you can look at #1091 on how do add the even details.

**Comment by dtacalau** [comment]
> I was thinking about this a bit and let's do this only when providing `{:#?}`, normally this means the >output is pretty printed, but we can (ab)use to also add additional platforms specific information. >For Unix you can look at #1091 on how do add the even details.

@Thomasdezeeuw could you please give more details, ATM is not clear to me what would you like pretty-printed for each platform. Thanks!

**Comment by Thomasdezeeuw** [comment]
@dtacalau sorry I wasn't really clear. What I meant was checking the [`fmt.alternate()` flag](https://doc.rust-lang.org/std/fmt/struct.Formatter.html#method.alternate) and to only show the platform specific details then. This way regular users of Mio won't be surprised/confused by them.

#1091, specifically https://github.com/tokio-rs/mio/pull/1091/files#diff-8be43826e3d9aa67792b01b56899deb9R331-R640, has the code on how to format the platform specific event data.

That would result in something like:

```rust
// `event` is readable & writable.
println!("{:?}", event); // Prints: Event { token: Token(0), readable: true, writable: true ... etc. } (no change).
println!("{:#?}", event); // Prints: Event { (same fields as above) .. platform: EVFILT_READ | EVFILT_WRITE }
```

**Comment by dtacalau** [comment]
@Thomasdezeeuw please review, thanks!

File: src/event/event.rs [hunk 1]
```diff
@@ -190,16 +190,24 @@ impl Event {
 
 impl fmt::Debug for Event {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Event")
-            .field("token", &self.token())
+        let alternate = f.alternate();
+        let d = &mut f.debug_struct("Event");
```
Comment (Thomasdezeeuw) [inline src/event/event.rs]: Why `&mut f.debug_struct("Event")`?

File: src/sys/unix/selector/epoll.rs [hunk 2]
```diff
@@ -121,6 +121,7 @@ pub type Events = Vec<Event>;
 pub mod event {
     use crate::sys::Event;
     use crate::Token;
+    use std::fmt::Write as FmtWrite;
```
Comment (Thomasdezeeuw) [inline src/sys/unix/selector/epoll.rs]: You're just import the trait right, so there is no need for the `as FmtWrite`.

Comment (Thomasdezeeuw) [reply src/sys/unix/selector/epoll.rs]: Same for the other places.

File: src/event/events.rs [hunk 3]
```diff
@@ -210,8 +210,13 @@ impl<'a> Iterator for Iter<'a> {
...
 impl fmt::Debug for Events {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Events")
-            .field("capacity", &self.capacity())
-            .finish()
+        let mut d = f.debug_struct("Events");
+        d.field("capacity", &self.capacity());
+
+        for event in self.iter() {
+            d.field("event", event);
```
Comment (Thomasdezeeuw) [inline src/event/events.rs]: Now we get a bunch of "event" which seems odd. Not sure about a better alternative though.

File: src/event/events.rs [hunk 4]
```diff
@@ -210,8 +210,13 @@ impl<'a> Iterator for Iter<'a> {
 
 impl fmt::Debug for Events {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Events")
-            .field("capacity", &self.capacity())
-            .finish()
+        let mut d = f.debug_struct("Events");
+        d.field("capacity", &self.capacity());
+
+        for event in self.iter() {
```
Comment (Thomasdezeeuw) [inline src/event/events.rs]: Let's only do this when alternative is enabled, just like with `Event`.

File: src/event/event.rs [hunk 5]
```diff
@@ -190,16 +190,24 @@ impl Event {
...
             .field("read_closed", &self.is_read_closed())
             .field("write_closed", &self.is_write_closed())
             .field("priority", &self.is_priority())
             .field("aio", &self.is_aio())
-            .field("lio", &self.is_lio())
-            .finish()
+            .field("lio", &self.is_lio());
+
+        if alternate {
+            let mut data_str = String::new();
```
Comment (Thomasdezeeuw) [inline src/event/event.rs]: Can we find a better way to do this then allocating a string?

File: src/event/event.rs [hunk 6]
```diff
@@ -190,16 +190,24 @@ impl Event {
...
+            sys::event::get_platform_event_data_str(&mut data_str, &self.inner);
```
Comment (Thomasdezeeuw) [inline src/event/event.rs]: Java level function name ;)

File: src/sys/windows/event.rs [hunk 7]
```diff
@@ -49,6 +51,37 @@ pub fn is_lio(_: &Event) -> bool {
...
+        POLL_SEND,
+        POLL_DISCONNECT,
+        POLL_ABORT,
+        POLL_LOCAL_CLOSE,
+        POLL_ACCEPT,
+        POLL_CONNECT_FAIL,
+    );
+
+    write!(data_str, ")").unwrap();
+    ()
```
Comment (Thomasdezeeuw) [inline src/sys/windows/event.rs]: Not needed.

File: tests/tcp_listener.rs [hunk 8]
```diff
@@ -343,6 +343,34 @@ fn try_clone_different_poll() {
     thread_handle2.join().expect("unable to join thread");
 }
 
+#[test]
+fn platform_specific_data_print() {
```
Comment (Thomasdezeeuw) [inline tests/tcp_listener.rs]: I don't really think there is a need for a test as the output is not stable.

If you do want a test I would suggest making one per platform the there `event` module and manually creating the `Event`s.

File: src/event/event.rs (hunk 5 above)
Comment (Thomasdezeeuw) [reply src/event/event.rs]: Would be better if we can somehow pass `fmt::Formatter`.

File: src/event/event.rs (hunk 6 above)
Comment (Thomasdezeeuw) [reply src/event/event.rs]: Maybe call it `details`?

File: src/sys/unix/selector/epoll.rs [hunk 9]
```diff
@@ -168,6 +169,46 @@ pub mod event {
         // Not supported.
         false
     }
+
+    pub fn get_platform_event_data_str(data_str: &mut String, event: &Event) {
+        write!(data_str, "events (").unwrap();
```
Comment (Thomasdezeeuw) [inline src/sys/unix/selector/epoll.rs]: This is not needed in the `Debug` impl the field name is already added.

File: src/sys/unix/selector/epoll.rs (hunk 2 above)
Comment (dtacalau) [reply src/sys/unix/selector/epoll.rs]: ok, I removed "as FmtWrite"

File: src/event/event.rs (hunk 1 above)
Comment (dtacalau) [reply src/event/event.rs]: Removed &

File: src/event/events.rs (hunk 3 above)
Comment (dtacalau) [reply src/event/events.rs]: Removed field name "event", was printed twice.

File: src/event/events.rs (hunk 4 above)
Comment (dtacalau) [reply src/event/events.rs]: done

File: src/event/event.rs (hunk 5 above)
Comment (dtacalau) [reply src/event/event.rs]: Redone by passing the fmt::Formatter.

File: src/event/event.rs (hunk 6 above)
Comment (dtacalau) [reply src/event/event.rs]: renamed to write_details, fwiw I like long explanatory names :)

File: src/sys/windows/event.rs (hunk 7 above)
Comment (dtacalau) [reply src/sys/windows/event.rs]: probably, removed!

File: src/sys/unix/selector/epoll.rs (hunk 9 above)
Comment (dtacalau) [reply src/sys/unix/selector/epoll.rs]: Removed

File: tests/tcp_listener.rs (hunk 8 above)
Comment (dtacalau) [reply tests/tcp_listener.rs]: Removed the test, not worth the trouble atm.

**Comment by dtacalau** [comment]
Here's how details are printed after changes:

- w/o alternate
```
poll events: Events { capacity: 16 }
```
- w/ alternate
```
poll events: Events {
    capacity: 16,
    : Event { token: Token(1), readable: false, writable: true, error: false, read_closed: false, write_closed: false, priority: false, aio: false, lio: false, platform: POLL_SEND },
}
```

**Pull request closed without merging**